fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/impl/LuceneSearchProvider.java

```diff
@@ -597,6 +591,11 @@ private Query getFilterQuery(Set<ApiFilter> filters) {
 
         lock.readLock().lock();
         try {
+            // return from cache if available
+            SinglePagePagination<DimensionRow> page = queryCache.get(new Pair<>(query, paginationParameters));
```

Comment:
This bothers me. You have an operation that _looks_ like a read inside a readers lock, but since this cache is actually mutating internal state as it handles the LRU magic, it may actually need a writers lock to avoid corrupting itself.